HashStore 1.1.0 Release

doulikecookiedough opened the pull request:
**Summary of Changes:**
- Major refactor of HashStore involving storing objects with their content identifiers `cid` and creating reference files
    - Storing an object with a pid will move the object to where it belongs, create a `pid` reference file that contains the `cid`, and a `cid` reference file that contains `pid`s that reference a `cid`
    - Storing an object without a pid will only store the object. The caller/client must call `tag_object` to create the reference files (else create an orphaned object), and optionally `delete_if_invalid_object` should they have data they want to validate the `object_metadata` returned from storing an object.
- Metadata is now stored using a object named with the hash of the `pid+format_id` and in a directory formed using the hash of the pid in `/metadata`
- New Public API Methods
    - `tag_object`, `delete_if_invalid_object`
- Improved thread safety and synchronization to address possible race conditions when working with data objects and added thread safety when working with metadata.
- The `hashstore` config file is now separated in two pieces, with the actual .yaml content being formed with a library
- HashStore Client Improvements
    - New script entry point after installing hashstore through `poetry install`
- Revised python docstrings into reStructuredText, a format compatible with `sphinx-autodocumentation` along with typehints
- New functionality to use python threads or processes via setting an environment variable
- Revised `README.md` to provide context to the HashStore library and how-to sections.
- Overall code clean up, revision to logging statements and various minor bug fixes

Greetings @artntek - When you have scope, can you please help review `HashStore` to assist me in releasing `HashStore 1.1.0`? Please do not hesitate to scrutinize any of the code or bring to my attention any concerns. Thank you 🙏!

artntek commented on src/hashstore/filehashstore.py:
```diff
@@ -1021,80 +1460,241 @@ def _mktmpmetadata(self, stream):
+        :rtype: str
+        """
+        checked_format_id = None
+        if format_id is not None and format_id.replace(" ", "") == "":
```
does this work as a more concise/readable option?

```python
if format_id and not format_id.strip():
```

doulikecookiedough replied:
Beautiful.... thank you!

artntek commented on src/hashstore/hashstore.py:
```diff
@@ -211,13 +255,16 @@ def get_hashstore(module_name, class_name, properties=None):
-        obj_size (bytes): Size of the object
-        hex_digests (dict, optional): A list of hex digests to validate objects
-            (md5, sha1, sha256, sha384, sha512)
+    :param str id: A unique identifier for the object (Hash ID, hex digest).
```
incorrect parameter name in docs

artntek commented on src/hashstore/hashstoreclient.py:
```diff
@@ -234,11 +245,11 @@ class HashStoreClient:
-        Args:
-            properties: See FileHashStore for dictionary example
-            testflag (str): "knbvm" to initialize MetacatDB
+        :param str origin_dir: Directory to convert.
+        :param str obj_type: Type of objects ('object' or 'metadata').
+        :param int num: Number of files to store.
```
none of these params match the method signature

doulikecookiedough replied:
D'oh! I have revised the doc string.

artntek commented on src/hashstore/hashstoreclient.py:
```diff
@@ -189,12 +200,12 @@ def __init__(self):
     def load_store_properties(self, hashstore_yaml):
```
method can be static

doulikecookiedough replied:
This is completed, thank you!

artntek commented on tests/test_filehashstore.py:
general question: my understanding is that methods starting with an underscore are "private", or used only by the class itself? If so, shouldn't all the methods that are NOT defined in the hashstore interface start with an underscore? Or am I interpreting this incorrectly?

doulikecookiedough replied:
AFAIK, access modifiers do not actually exist in Python. Using an underscore implies that it is "private" and not meant to be used publicly - I have been a little loose with where I add _ and don't add _, with the criteria being would this be helpful for the client to be able to do or not. To reduce confusion here, I will add an underscore to all non-Public API methods.

artntek commented on src/hashstore/filehashstore.py:
```diff
@@ -404,77 +412,240 @@ def lookup_algo(algo):
+                raise FileExistsError(exception_string)
+            elif os.path.exists(cid_ref_abs_path):
+                # Create the pid refs file
+                pid_tmp_file_path = self._write_pid_refs_file(tmp_root_path, cid)
+                self.create_path(os.path.dirname(pid_ref_abs_path))
+                shutil.move(pid_tmp_file_path, pid_ref_abs_path)
```
curious why you create a temp file then move it? The pid files are tiny, right?

doulikecookiedough replied:
I followed this pattern so that this process of writing the refs file could be atomic (either the file was written successfully or it's not written at all, and they are tiny indeed!).

artntek commented on src/hashstore/filehashstore.py:
```diff
@@ -404,77 +412,240 @@ def lookup_algo(algo):
+                )
+                return True
+            else:
+                # All ref files begin as tmp files and get moved sequentially at once
```
same question as above

doulikecookiedough replied:
^ Similar to above, I wanted to have all the files prepped/ready so that I could move them together at once to reduce the chances of a missing refs file - so both `cid` and `pid` refs file are moved together to where they need to be at the same time.

artntek reviewed (changes requested):
see inline comments

doulikecookiedough commented on README.md:
```diff
@@ -54,49 +67,161 @@ my_store = factory.get_hashstore(module_name, class_name, properties)
+
+**How do I find an object or check that it exists if I have the pid?**
+- To find the location of the object, call the Public API method `find_object` which will return the content identifier (cid) of the object.
+- This cid can then be used to locate the object on disk by following HashStore's store configuration.
```
Initially, I wanted to have an "exists" method - which led to me using `find_object` for more than one purpose as I had documented above. In the worst case scenario where someone would have to go in and manually fix things, I thought this would be a helpful method... Only someone with the right privileges would be able to make use of this actual location. I would like to keep this in, but if you feel strongly that it shouldn't please let me know and let's discuss.

artntek replied:
Exposing the details of internal implementations to external clients is a fundamental design flaw, not just my preference :-)  https://en.wikipedia.org/wiki/Encapsulation_(computer_programming)

I have seen it lead to many serious situations IRL, where client code relies on these internal details, and then their code breaks when those details change.

If you think this method is needed, then it should not be be part of a public API, but should instead have some form of access restriction so it's usable only by admins, for example

doulikecookiedough replied:
Note - this has been resolved (we have removed this from the Public API)

doulikecookiedough commented on README.md:
```diff
@@ -54,49 +67,161 @@ my_store = factory.get_hashstore(module_name, class_name, properties)
+
+In HashStore, objects are first saved as temporary files while their content identifiers are calculated. Once the default hash algorithm list and their hashes are generated, objects are stored in their permanent location using the store's algorithm's corresponding hash value, the store depth and the store width. Lastly, reference files are created for the object so that they can be found and retr ...
+
+By calling the various interface methods for  `store_object`, the calling app/client can validate, store and tag an object simultaneously if the relevant data is available. In the absence of an identfiier (ex. persistent identifier (pid)), `store_object` can be called to solely store an object. The client is then expected to call `verify_object` when the relevant metadata is available to confirm  ...
```
I discussed `verify_object` with Jing and we determined that it should return a boolean value. If false is returned, the client is expected to delete the object directly. The ability to do this will be present once I refactor `delete_object` based on the new requirements described here: [Refactor deleteObject for new requirements](https://github.com/DataONEorg/hashstore-java/issues/57)

To your second question - yes, we will end up with a bunch of orphaned objects if a client calls `store_object` and forgets to `tag_object` (technically, you don't have to call `verify_object`...)

doulikecookiedough commented on README.md:
````diff
@@ -54,49 +67,161 @@ my_store = factory.get_hashstore(module_name, class_name, properties)
+
+###### What does HashStore look like?
+
 ```
````
Thank you!

doulikecookiedough commented on src/hashstore/filehashstore.py:
```diff
@@ -404,77 +412,240 @@ def lookup_algo(algo):
+                    "FileHashStore - store_object: %s is currently being stored. Waiting.",
+                    pid,
+                )
+                time.sleep(self.time_out_sec)
```
Yes, your assumption is correct.

doulikecookiedough commented:
Thanks again @artntek for your review! I've made the updates in another branch and merged them into develop. I'm also going to hold off on the 1.1.0 release until the changes to `deleteObject` are implemented (and merged into develop). I will ping you again then to wrap this PR up.

doulikecookiedough commented:
Greetings @jeanetteclark! When you have scope, can you please help review/perform a sanity check on HashStore?

**Additional Changes Since Matthew's Review:**
- Refactored `delete_object` functionality & scope
    - Clients can delete a single object (based on a given content identifier `cid`)
    - Clients can delete an object and all of its related metadata when given a pid
- Optimized `tag_object` flow with additional locks
    - Added custom exception classes and revised logging statements to improve clarity 
- HashStore Client Updates
    - Revised the way in which the client is called via an executable installed by poetry
    - Fixed bug in storing metadata (incorrect name due to changes not accounted for)
- Improvements to synchronization process via usage of `Conditions`
    - Added multiprocessing locks & conditions
    - _Notable Change: Test times executing in ~2s vs ~15s previously_
- Updated README.md

jeanetteclark reviewed (approved):
Things look good - I created a local store using the client and all my testing there did great. I didn't go into any great detail into the code but I made one minor comment on something to improve. I don't think its a make or break thing though so I'm going to approve.

doulikecookiedough commented on src/hashstore/hashstoreclient.py:
```diff
@@ -585,21 +610,31 @@ def get_object_metadata_list(self, origin_directory, num):
+                continue
+            else:
+                # Get pid, filepath and formatId
+                pid_guid = row[0]
```
Thank you for your feedback @jeanetteclark! When I receive the results from the SQL query to retrieve specific columns/values - the column names are not included in the query result set (each row is a tuple item). The only way to access each row would be through the index positions. 

For now, I am going to leave this part as is because it doesn't appear to be a straightforward fix. Should the db schema change, my SQL query will also need to be updated. I will review this section more closely when that happens.

jeanetteclark replied:
sounds good - I wasn't sure what the format would look like on return. Definitely not a critical fix

artntek reviewed (dismissed):
See slack - not sure if this is even still relevant :-)
